runtime/src/androidTest/kotlin/pbandk/internal/Util.kt

```diff
@@ -0,0 +1,12 @@
+package pbandk.internal
+
+import java.util.*
+
+/**
+ * Same implementation as JVM to run unit tests on Android as [android.util.Base64] will throw
+ * a NotMockedException.
```

Comment:
Hmm. The downside of this approach is that the unit tests don't test the actual Android implementation. For example, it's possible that Android's `Base64` might handle edge cases (such as support for Base64Url, line wrapping, etc.) differently from Java's `Base64`. But there's no way to verify that by running the unit tests 😞 

Unfortunately I don't have any great suggestions here. The only thing I can think of is to run the unit tests on Android using Roboelectric. But from some initial googling, I don't see how we can configure the Kotlin Multiplatform tests that come from `commonTest` to use the Roboelectric test runner for the `android` platform.

Reply:
I think I got something that works! Add Roboelectric's Android stubs as a runtime test dependency:

```kt
        val androidTest by getting {
            dependencies {
                ...
                runtimeOnly("org.robolectric:android-all:11-robolectric-6757853")
            }
        }
```

Then delete the `androidTest` version of `Util.kt`. The tests should now run correctly with the Android Base64 library. There's one test with a legitimate failure that I'll leave up to you to fix 😉

Reply:
If I use Robolectric `v11-...` then I get the error:

```
java.lang.UnsupportedClassVersionError: android/util/Base64 has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

And if I use Robolectric `v10-...` (october 2019) I get the error: 

```
java.lang.AssertionError: expected:<{"optionalBytes":"AQI="}> but was:<{"optionalBytes":"AQI=\n"}>
```

Two thoughts about the failure:

- The version of `Robolectric` used is already almost 1.5 years old (installing a newer version of the Java SDK might solve this though)
- Is this failure representative for Android or is this a bug in `Robolectric`

--------------------------------------

Few thoughts on using Robolectric:

1. The platform specific `Base64` implementation wan't tested before
before adding the Android artifact, we were using the Android `java.util.Base64` which was already a different implementation from the JDK `java.util.Base64`

2. Robolectric is only an approximation of the Android framework
imagine that the tests pass using Robolectric, how much guarantee does that give for an actual Android device? In the end Robolectric has some version of `android.util.Base64` hardcoded into it, potentially even modified so it works run during unit tests.

3. What value would such a test add?
even if the `android.util.Base64` implementation of Robolectric is exactly the same as the one in the Android platform, then this test only serves to prove that `android.util.Base64` works. 
Shouldn't that be the responsibility of the Android team? Note also that this class exists since API level 1, has a proven track record and has only been [modified 6 times the past 10 years](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/util/Base64.java;l=29;bpv=1). 
We aren't going to test Android's String implementation either, right? (in this regard I highly recommend this article: https://www.facebook.com/notes/kent-beck/unit-tests/1726369154062608/)

------------------------
So while everything can be tested, not all tests are created equal!
Sometimes the value added isn't proportional to the effort it took writing and maintaining such a test.

To me personally Robolectric is one such case: I've honestly completely given up on that framework. It slows my tests down, is a pain to maintain and most important of all it doesn't spark joy.

-----------------------
That being said, if we choose to use Robolectric, I would highly appreciate your advice on to fix:

```
java.lang.AssertionError: expected:<{"optionalBytes":"AQI="}> but was:<{"optionalBytes":"AQI=\n"}>
```

Reply:
The actual implementation would be a bit hairier than this as `Build.VERSION.SDK_INT` is 0 during unit tests:

```kotlin
actual override fun base64ToBytes(str: String): ByteArray =
    if (doesntSupportJavaBase64()) {
        android.util.Base64.decode(str, android.util.Base64.DEFAULT)
    } else {
        java.util.Base64.getDecoder().decode(str)
    }
    
// Also need to check for minimum version as [SDK_INT] is 0 during unit tests
    private inline fun doesntSupportJavaBase64() =
        Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP && Build.VERSION.SDK_INT < Build.VERSION_CODES.O
  ```
   
Begs the question that try/catching wouldn't be better in this case:

```kotlin
actual override fun base64ToBytes(str: String): ByteArray = try {
    java.util.Base64.getDecoder().decode(str)
} catch (e: Throwable) {
    // Fallback to Android implementation on platforms below API 26
    android.util.Base64.decode(str, android.util.Base64.DEFAULT)
}
```

Reply:
I've added such an implementation, not to push this either way, but just to show you how that would look like.

Please advice on how you prefer to proceed.

Reply:
I should've been more specific in my comment yesterday. This:

```
java.lang.AssertionError: expected:<{"optionalBytes":"AQI="}> but was:<{"optionalBytes":"AQI=\n"}>
```

is an actual bug in pbandk that was found by running the tests with Robolectric. It's a bug that would've occurred on a real Android device. The purpose of running unit tests with Robolectric is not to guard against bugs in Android's implementation of standard Java APIs. I agree with you that that would be overkill and completely unnecessary. We can assume Android's implementation is well-tested. The purpose of running unit tests with Robolectric is to guard against bugs where pbandk's `android` variant uses different APIs from the `jvm` variant (such as with Base64), which might work subtly differently. Adding an API check to the code, as you're suggesting, doesn't fix the problem. You'll still have a bug on devices with an API level below `Build.VERSION_CODES.O`.

It looks like the cause of this bug is because `android.util.Base64` by default line wraps the Base64 output. Whereas `java.util.Base64` does not. No line wrapping is the behavior that other protobuf libraries will expect. I think changing the code to this will fix the bug:

```kt
Base64.encodeToString(bytes, Base64.NO_WRAP)
```

Reply:
Regarding the error you get when using the `v11` Robolectric library:

```
java.lang.UnsupportedClassVersionError: android/util/Base64 has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

I suspect this is because you're running the tests with Java 8 on your local machine (I think this is what Android Studio bundles). Whereas it looks like the Robolectric `v11` library was compiled with Java 9. I'm running the tests under Java 15 on my machine, which is why I didn't hit this issue. I'm fine with using `v10` for now to maintain Java 8 compatibility. I think Android is planning to update to Java 11 in the next Android Studio release.